proposals/NNNN-hashable-enhancements.md

```diff
@@ -0,0 +1,925 @@
+[MurmurHash]: https://github.com/aappleby/smhasher
+[CityHash]: https://github.com/google/cityhash
+[SipHash]: https://131002.net/siphash/
+[HighwayHash]: https://github.com/google/highwayhash
+
+Since [SE-0185] required the standard library to implement a
+high-quality universal hash function, it seems like a good idea to
+expose it as public API, so that manual `Hashable` implementations can
+take advantage of it, too.
+
+Universal hash functions work by maintaining some internal state --
+this can be as simple as a single 32/64-bit integer value (for
+e.g. [FNV-1a]), but it is usually much wider than that. For example,
+[SipHash] maintains a state of 256 bits, while [HighwayHash] uses 1024
+bits. 
+
+## <a name="proposed-solution">Proposed solution</a>
+
+We solve `Hashable`'s implementation problems in two parts. First, we
+make the standard library's hash function public. Second, we replace
+`hashValue` with a requirement that is designed specifically to
+eliminate the guesswork from manual `Hashable` implementations.
+
+### <a name="hasher">The `Hasher` struct</a>
+
+We propose to expose the standard library's standard hash function as
+a new, public struct type, called `Hasher`. This new struct captures
+the state of the hash function, and provides the following operations:
+
+1. An initializer to create an empty state. To make hash values less
+   predictable, the standard hash function uses an arbitrary 128-bit
+   seed value, specified as a tuple of two `UInt64` integers.
+   `Hasher` uses a per-execution random seed value by default;
+   however, we also allow programmers to specify their own seed
+   values, primarily to enable per-instance random seeding in their
+   custom hashing collections.
+   
+    ```swift
+    public struct Hasher {
+      public init()         // Use the default per-execution random seed value
+      public init(seed: (UInt64, UInt64)) // Use the specified seed value
+    }
+    ```
+
+2. A set of operations to mix in new bits into the state of the hash
+   function. For reasons of efficiency, these are built around
+   appending integer values to the hasher:
+   
+    ```swift
+    extension Hasher {
+      public mutating func append(bits: Int)
+      public mutating func append(bits: UInt)
+      public mutating func append(bits: Int64)
+      public mutating func append(bits: UInt64)
+      public mutating func append(bits: Int32)
+      public mutating func append(bits: UInt32)
+      // etc.
+    }
+    ```
+
+    (We expect most hashable types will have discrete components; the
+    bits to be hashed will rarely be available as a contiguous
+    byte sequence.)
+
+3. An operation to finalize the state, extracting the hash value from it.
+   
+    ```swift
+    extension Hasher {
+      public mutating func finalize() -> Int
+    }
+    ```
+
+   Finalizing the hasher state invalidates it; it is illegal to call
+   `append` or `finalized` on a finalized hasher.
```

Comment:
BTW, do we enforce this? At the very least, with a debugPrecondition?

Reply:
Not yet, but we will soon. `_Hasher` currently misses half of the API proposed here; I'll need to open a new PR in the swift repo to add the missing pieces and de-underscore it. We can use the an unused bit pattern in the fifth state word to mark finalized hashers.